arangod/Cluster/MaintenanceFeature.cpp

```diff
@@ -921,6 +921,35 @@ uint64_t MaintenanceFeature::shardVersion(std::string const& shname) const {
 
 uint64_t MaintenanceFeature::incShardVersion(std::string const& shname) {
   MUTEX_LOCKER(guard, _versionLock);
+
+  uint64_t v = 0;
+  using namespace std::chrono;
+  using clock = steady_clock;
+  auto timeout = duration<double>(600.0);
+  auto stoppage = clock::now();
+  auto snooze = milliseconds(100);
+
+  // Acquire current version from agency and wait for it to have been dealt
+  // with in local current cache. Any future current version will do, as
+  // the version is incremented by the leader ahead of getting here on the
+  // follower.
+  while (!server().isStopping() && clock::now() < stoppage) {
+    cluster::fetchCurrentVersion(0.1 * timeout)
+      .thenValue(
+        [&v] (auto&& res) { v = res.get(); })
+      .thenError<std::exception>(
+        [&shname] (std::exception const& e) {
+          LOG_TOPIC("3ae99", ERR, Logger::CLUSTER)
+            << "Failed to acquire current version from agency while increasing shard version: "
+            << " for shard "  << shname << e.what();
+        });
+    std::this_thread::sleep_for(snooze);
+    if(snooze < seconds(2)) {
+      snooze += milliseconds(100);
+    }
+  }
+  server().getFeature<ClusterFeature>().clusterInfo().waitForCurrentVersion(v).wait();
```

Comment:
Actually, this line looks like the right solution. Is there any reason why we need the above loop at all instead of just the single request and wait?

Reply:
If we get a shutdown in the meantime, we should get a connection canceled and an error on the future.

Reply:
But Max is right, we do need to actually handle the value/error here.

Reply:
we need to know the current version in ClusterInfo, not only agency cache is beyond the time we got the repl done info